packages/contracts-bedrock/test/cannon/MIPS2.t.sol

```diff
@@ -0,0 +1,1787 @@
+import { MIPSSyscalls as sys } from "src/cannon/libraries/MIPSSyscalls.sol";
+import "src/dispute/lib/Types.sol";
+
+contract ThreadStack {
+    bytes32 internal constant EMPTY_THREAD_ROOT = hex"ad3228b676f7d3cd4284a5443f17f1962b36e491b30a40b2405849e597ba5fb5";
+
+    struct Entry {
+        MIPS2.ThreadState thread;
+        bytes32 root;
+    }
+
+    Entry[] internal stack;
+
+    constructor() {
+        stack.push();
+        stack[stack.length - 1].root = EMPTY_THREAD_ROOT;
+    }
+
+    function root() public view returns (bytes32) {
+        return stack[stack.length - 1].root;
+    }
+
+    function inner(uint256 _i) public view returns (bytes32 root_) {
+        root_ = stack[stack.length - 1 - _i].root;
+    }
+
+    function top() public view returns (MIPS2.ThreadState memory thread_) {
+        thread_ = peek(0);
+    }
+
+    function peek(uint256 _i) public view returns (MIPS2.ThreadState memory thread_) {
+        thread_ = stack[stack.length - 1 - _i].thread;
+    }
+
+    function push(MIPS2.ThreadState memory _thread) public {
+        _push(_thread);
+    }
+
+    function pop() public {
+        stack.pop();
+    }
+
+    function replace(MIPS2.ThreadState memory _thread) public {
+        stack.pop();
+        _push(_thread);
+    }
+
+    function _push(MIPS2.ThreadState memory _thread) internal {
+        bytes32 newRoot = keccak256(abi.encodePacked(stack[stack.length - 1].root, keccak256(encodeThread(_thread))));
+        stack.push(Entry(_thread, newRoot));
+    }
+}
+
+contract Threading {
+    ThreadStack public left;
+    ThreadStack public right;
+    bool public traverseRight;
+    uint32 public nextThreadID;
+
+    constructor() {
+        left = new ThreadStack();
+        right = new ThreadStack();
+        traverseRight = false;
+    }
+
+    function createThread() public returns (MIPS2.ThreadState memory thread_) {
+        thread_.threadID = nextThreadID;
+        if (traverseRight) {
+            right.push(thread_);
+        } else {
+            left.push(thread_);
+        }
+        nextThreadID += 1;
+    }
+
+    function current() public view returns (MIPS2.ThreadState memory out_) {
+        if (traverseRight) {
+            out_ = right.top();
+        } else {
+            out_ = left.top();
+        }
+    }
+
+    function replaceCurrent(MIPS2.ThreadState memory _thread) public {
+        if (traverseRight) {
+            right.replace(_thread);
+        } else {
+            left.replace(_thread);
+        }
+    }
+
+    function witness() public view returns (bytes memory out_) {
+        if (traverseRight) {
+            out_ = abi.encodePacked(encodeThread(right.top()), right.inner(1));
+        } else {
+            out_ = abi.encodePacked(encodeThread(left.top()), left.inner(1));
+        }
+    }
+
+    function setTraverseRight(bool _traverseRight) public {
+        traverseRight = _traverseRight;
+    }
+}
+
+contract MIPS2_Test is CommonTest {
```

Comment:
Nice test coverage!  I can think of a few more cases we might want to cover:
- I don't think there are any tests that check that we wakeup sleeping threads when `State.wakeup` is not set
- Might be a good idea to have a few tests that set up different numbers of threads, and walk through them over multiple steps in wakeup and regular modes of operation to check our thread iteration works

Reply:
Agreed. I'll add them.

Reply:
Added a couple tests for single direction traversals. We should add more traversal scenarios in the Go differential tests later. It's easier to do this in Go w/ hand written assembly than in solidity.